dotnet/src/SemanticKernel.Abstractions/Functions/KernelArguments.cs

```diff
@@ -65,7 +86,34 @@ public KernelArguments(IDictionary<string, object?> source, Dictionary<string, P
     /// <summary>
     /// Gets or sets the prompt execution settings.
     /// </summary>
-    public IReadOnlyDictionary<string, PromptExecutionSettings>? ExecutionSettings { get; set; }
+    public IReadOnlyDictionary<string, PromptExecutionSettings>? ExecutionSettings
+    {
+        get => this._executionSettings;
+        set
+        {
+            this._executionSettings = value;
+
+            if (this._executionSettings is null ||
+                this._executionSettings.Count == 0)
+            {
+                return;
+            }
+
+            foreach (var kv in this._executionSettings)
+            {
+                // Ensures that if a service id is not specified and is not default, it is set to the current service id.
+                if (kv.Key != kv.Value.ServiceId)
+                {
+                    if (!string.IsNullOrWhiteSpace(kv.Value.ServiceId))
+                    {
+                        throw new ArgumentException($"Service id '{kv.Value.ServiceId}' must match the key '{kv.Key}'.", nameof(this.ExecutionSettings));
+                    }
+
+                    kv.Value.ServiceId = kv.Key;
+                }
+            }
```

Comment:
What happens if someone sets this value and then mutates the Dictionary later? Should we create a copy to prevent this?

Reply:
Yeah, I did this in the PromptExecutionsettings and skipped this, good catch.